Add missing offline access/refresh token doc from GSMA (#123)
camaraproject/IdentityAndConsentManagement: documentation/CAMARA-API-access-and-user-consent.md

More details about the standard flow can be found in the official IETF specification [The OAuth 2.0 Authorization Framework](https://datatracker.ietf.org/doc/html/rfc6749#section-4.4).

## Offline Access - Refresh Token

> The defined solution uses Refesh Token as a mean to enable CAMARA API access in Offline Scenarios. Online Access means that there is a Telco-based Network Connection, either Mobile Network connection in mobile use cases or Fixed Network connection in fixed use cases.

Reviewer: @AxelNennker Could you suggest a clearer way to describe what constitutes an online scenario (the telco is able to identify the connected user) as opposed to offline scenarios (e.g., when the user connects to the Internet via WiFi in a coffee shop)?

> The defined solution uses Refesh Token as a mean to enable CAMARA API access in Offline Scenarios. Online Access means that there is a Telco-based Network Connection, either Mobile Network connection in mobile use cases or Fixed Network connection in fixed use cases.

Reviewer 2: When the user logs in and gives their consent then the use does not have to be "on-net".

Nit: the telco is in most cases not identifying the "user" but the subscriber.

If using methods .like operator token as defined in section 14 of https://www.gsma.com/newsroom/wp-content/uploads//TS.43-v10.0-Service-Entitlement-Configuration.pdf the device does not have to be "on-net" to identify the subscriber.

I'll try to find a description that fits here.

- Refresh Token may be issued as result of AuthN/AuthZ, if requested by API invoker and allowed by use case. 
  - [RFC6749 Section 1.5](https://datatracker.ietf.org/doc/html/rfc6749#section-1.5): _A refresh token is a string representing the authorization granted to  the client by the resource owner.  The string is usually opaque to the client. The token denotes an identifier used to retrieve the authorization information. Therefore, the Refresh Token is bound to the authorization of the end user to the Application._

> - As indicated by [OIDC Spec](https://openid.net/specs/openid-connect-core-1_0.html#OfflineAccess ), one of the uses of refresh token is allowing offline access. 

Reviewer 3: I think this section is conflating things.

The `offline_access` scope is specifically defined in OpenID Connect to mean offline access to 'End-User's UserInfo Endpoint.'

This means when it receives an offline_access request, the authorization server can ask the user if they wish to grant ongoing access to the userinfo.

This same `offline_access` scope should not be extended to have the meaning of requesting equivalent ongoing access to other APIs. If there are other APIs that need to distinguish whether short or long term access is being requested those APIs should define explicit scope values with those meanings.

Author: @jogu (and CC @AxelNennker) That's why this usage of the `offline_access` scope is included in the proposed CAMARA OIDC profile (https://github.com/camaraproject/IdentityAndConsentManagement/pull/113), which covers the upcoming GSMA Opengateway technical stream requirement for off-net access scenarios. For example, in scenarios where the device is no longer connected to the operator's network. This information was agreed and specified in the GSMA Opengateway technical stream, and as stated in the PR description, it was agreed to move it to CAMARA.

So this is indeed an extension of `offline_access` not only to the User Info endpoint, but also for Service APIs.

https://github.com/camaraproject/IdentityAndConsentManagement/blob/main/documentation/CAMARA-API-access-and-user-consent.md (release v0.1.0), containing CAMARA reached short-term agreements in this WG, also states the usage of `offline_access` for "Off-net" scenarios using refresh_token, as long as there was a connection when the first access_token was requested.

As recommended during today's WG meeting call (14/02), that we first focus on moving the necessary documentation to the CAMARA Identity & Consent WG repository as agreed. And then, if the WG decides so, we are more than happy to discuss this topic for release 0.3.0. focused on enhancements and features for the mid-term.

CC @trehman-gsma @MarkCornall

Reviewer 4: > And then, if the WG decides so, we are more than happy to discuss this topic for release 0.3.0. focused on enhancements and features for the mid-term.

@jpengar I agree that merging the PR and then raise an issue with the needed changes could be the agreed way forward. But if there is a security issue identified and/or the `offline_access` scope can't be extended to other APIs then this MUST be corrected before v0.2.0 can be released.

Author: @hdamker What is exactly the security issue you mean?

Reviewer 4: @jpengar nothing concrete as I'm not the expert here, hence the "if". I just wanted to emphasise that we can't postpone critical issues handle as enhancements and postpone to v0.3.0.

Reviewer: This was the agreed solution to **support the off-net scenarios**, e.g., I get an access token when I'm using my mobile connection, then I connect to a Starbucks WiFi and can use & refresh the token. Nobody argued it constitutes a security/privacy issue (let alone critical).

It was also agreed to move the docs from the private GSMA repository, to keep a **single source of truth**, because keeping two versions of the same docs was the worst-case scenario (see #82, https://github.com/camaraproject/IdentityAndConsentManagement/issues/70#issuecomment-1778852651, https://github.com/GSMA-Open-Gateway/Open-Gateway-Documents/issues/101, https://github.com/camaraproject/IdentityAndConsentManagement/blob/main/documentation/MeetingMinutes/2024/MOM-2024-01-17.md#redundancyoverlap-between-gsma-and-camara-doc-duplicated-discussions, etc).

Knowing that, and that  could we please merge it to get rid of the GSMA version, and then **discuss your alternatives to cover the off-net scenarios**?

Reviewer 2: I think the document should not call "it" offline_access and then talk about refresh tokens. That are different things. (As all know)
I still think that this is wrong and should not be merged.
Something being agreed upon at GSMA does not mean that sections identified now as wrong should be merged in Camara.
I prefer to fix wrongness first and then merge.

Reviewer: @AxelNennker I disagree for different reasons:
- There was discussion in the GSMA technical stream to find a solution to support off-net scenarios. This was the agreed solution. There was also an **agreement to move that doc from GSMA to CAMARA** to have a single source of truth. This is what this PR does, nothing more.
- **"Wrongness" is an opinion**. There are more solutions in the market implementing similar behaviors. For example, Okta, one of the most popular "identity as a service providers" for devs, does exactly the same
https://developer.okta.com/docs/guides/refresh-tokens/main/#get-a-refresh-token. Microsoft supports the same semantics https://learn.microsoft.com/en-us/entra/identity-platform/scopes-oidc#the-offline_access-scope. Atlassian uses offline_access exactly the same way https://developer.atlassian.com/cloud/oauth/getting-started/refresh-tokens/.  I can't judge whether this is one of those _"other contexts that are beyond the scope of this specification"_ that OIDC leaves the door open to, but it's definitely working well for thousands of devs.

I understand that you might not like it, since every solution has pros&cons (and of course we should keep improving it). But I don't think that's a good reason to block it.

Reviewer 3: >  I can't judge whether this is one of those "other contexts that are beyond the scope of this specification" that OIDC leaves the door open to, but it's definitely working well for thousands of devs.

It's helpful to quote the full paragraph so you have the context of those 'other contexts':

> The use of Refresh Tokens is not exclusive to the offline_access use case. The Authorization Server MAY grant Refresh Tokens in other contexts that are beyond the scope of this specification.

This paragraph is making it clear that refresh tokens can be issued even when scope=offline_access is not requested. It is not suggesting that offline_access should be used in other contexts.

For example in OpenData ecosystems, refresh tokens are very commonly used, but I have never seen one that requires the offline_access scope to get a refresh token - it is common for the length of access actually needed by the client application to be indicated to the authorization server though.

As you say, some people do appear to overload offline_access to mean other things. Some people have chosen to not allow applications to indicate which resources they need ongoing access to. However given the wide range of CAMRA APIs I am not sure a decision that a refresh token should only be requested by requesting offline_access is appropriate, and I might go as far as saying that more fine grained permissioning represents current best practice.

Reviewer 2: Is this text acceptable? I think there is no need to mention off-net or on-net. This section is taken from PR https://github.com/camaraproject/IdentityAndConsentManagement/pull/121

I guess the section should not be both in the security and interoperability profile and in CAMARA-API-access-and-user-consent. I think it should be in the "profile" and not in CAMARA-API-access-and-user-consent.

### Offline Access

Neither OIDC not OAuth2 define a way for clients to indicate whether they need a refresh_token. 
In OIDC and OAuth2 refresh token issuance is optional in Authorization Code Flow and at the discretion of the AZ.

Camara uses the scope `offline_access` in the authorization request to indicate to the AZ that the client requests a refresh token additionally to the access token for Camara API access.

---
**NOTE**

OIDC defines the scope `offline_access` and it is used to get an access token to be used at the UserInfo endpoint, and also to obtain a request token to refresh this access token.

---

The OIDC rules for `offline_access` as defined in [OIDC Offline Access](https://openid.net/specs/openid-connect-core-1_0.html#OfflineAccess) apply. 

Copied to here unchanged from the OIDC standard for clarity: 

>  When offline access is requested, a prompt parameter value of consent MUST be used unless other conditions for processing the request permitting offline access to the requested resources are in place. The OP MUST always obtain consent to returning a Refresh Token that enables offline access to the requested resources. A previously saved user consent is not always sufficient to grant offline access.

> Upon receipt of a scope parameter containing the offline_access value, the Authorization Server:

>    MUST ensure that the prompt parameter contains consent unless other conditions for processing the request permitting offline access to the requested resources are in place; unless one or both of these conditions are fulfilled, then it MUST ignore the offline_access request,

>    MUST ignore the offline_access request unless the Client is using a response_type value that would result in an Authorization Code being returned,

>    MUST explicitly receive or have consent for offline access when the registered application_type is web,

>    SHOULD explicitly receive or have consent for offline access when the registered application_type is native.

The OIDC section on [OIDC Using Refresh Tokens](https://openid.net/specs/openid-connect-core-1_0.html#RefreshTokens) is not changed by this document.

The OIDC security and privacy considerations regarding offline access and refresh tokens apply e.g. on
[Token lifetime](https://openid.net/specs/openid-connect-core-1_0.html#TokenLifetime) and [Offline Access Privacy](https://openid.net/specs/openid-connect-core-1_0.html#OfflineAccessPrivacy).

FE->>+FE: Browser /<br> Embedded Browser
alt Standard OIDC Auth Code Flow between Invoker and API Exposure Platform
  FE-->>ExpO: GET /authorize?response_type=code&client_id=coolApp<br>&scope=dpv:<purposeDpvValue>#35;<technicalParameter> offline_access<br>&redirect_uri=invoker_callback...
  rect rgb(0, 76, 153)
    Note over ExpO: Authorization process details included in<br>the "Consume a CAMARA API<br>- Authorization Code Grant (FrontEnd)" flow...
  end    
  ExpO-->>FE: 302<br>Location: invoker_callback?code=Operatorcode
  FE-->>-BE: GET invoker_callback?code=OperatorCode
  BE->>ExpO: POST /token<br> code=OperatorCode
  ExpO->>ExpO: Create and save<br>Operator Refresh Token
  ExpO->>BE: 200 OK <br> {OperatorAccessToken}<br>{OperatorRefreshToken}
end
```

**Key concepts**

- Same considerations that apply for general call flows in [Authorization flows / grant types](#authorization-flows--grant-types) also apply for this flow.
-  Refresh Token issuance is achieved by requesting `offline_access` scope. 

**Flow description** 

> The API invoker (which could potentially be the application backend, an aggregator, etc.) instructs the application frontend in the device to initiate the OIDC authorization code flow with the operator as described in the [Authorization code flow (Frontend flow)](#authorization-code-flow-frontend-flow).   

Reviewer 2: ```suggestion
The API invoker (which could potentially be the application backend, an aggregator, etc.) instructs the application frontend in the device to initiate the OIDC authorization code flow with the Authorization Server as described in the [Authorization code flow (Frontend flow)](#authorization-code-flow-frontend-flow).   
```

The API invoker includes `offline_access` scope to signal that a Refresh Token is expected (step 1). 

Then same flow depicted in [Authorization code flow (Frontend flow)](#authorization-code-flow-frontend-flow) occurs (steps 2-8). 

> **NOTE: Flow diagram is simplified, full details offered in general call flows** 

> When Operator is going to issue the Access Token, it also issues a Refresh Token (OperatorRefreshToken), as `offline_access` scope was requested. 

Reviewer 2: ```suggestion
When the Authorization Server is going to issue the Access Token, it also issues a Refresh Token, as `offline_access` scope was requested. 
```

Note over BE: Select User Identifier:<br> Ip:port / MSISDN / other TBD  

alt OIDC Client-Initiated Backchannel Authentication (CIBA) Standard Flow between Invoker and Operator.
  BE->>+ExpO: POST /bc-authorize<br> Credentials,<br>scope=dpv:<purposeDpvValue>#35;<technicalParameter>" offline_access,<br>login_hint including User Identifier    
  rect rgb(0, 76, 153)
    Note over ExpO: Authorization process details included in<br>the "Consume a CAMARA API - CIBA flow"...
  end  
  ExpO->>BE: HTTP 200 OK {"auth_req_id": "{OperatorAuthReqId}"
  loop Invoker polls until consent is granted or until expires. If granted in advance, token returned in first poll
    BE->>+ExpO: POST /token <br>Credentials}<br>auth_req_id={OperatorAuthReqId}
    ExpO->>ExpO: Create and save<br>Operator Refresh Token    
    ExpO->>-BE: HTTP 200 OK <br>{OperatorAccessToken}<br>{OperatorRefreshToken} 
  end  
end
```

**Key concepts**

- Same considerations that apply for general call flows in [Authorization flows / grant types](#authorization-flows--grant-types) also apply for this flow. 
- Refresh Token issuance is achieved by requesting `offline_access` scope 
 
**Flow description**

> The API invoker (which could potentially be the application backend, an aggregator, etc.) requests a 3-legged access token to the operator API exposure platform. The process follows the OpenID Connect Client-Initiated Backchannel Authentication (CIBA) flow as described in [CIBA flow (Backend flow)](#ciba-flow-backend-flow).  

Reviewer 2: ```suggestion
The API invoker (which could potentially be the application backend, an aggregator, etc.) requests a 3-legged access token to the API exposure platform. The process follows the OpenID Connect Client-Initiated Backchannel Authentication (CIBA) flow as described in [CIBA flow (Backend flow)](#ciba-flow-backend-flow).  
```

  end
  alt If Consent is NOT granted (consent expired/revoked) - The flow fails
    ExpO->>BE: 400 Bad Request<br>{"error":"invalid_scope"}        
  else If Consent is Granted or Consent not needed for legal basis   
    ExpO->>ExpO: Generate<br>new {NewOperatorAccessToken}<br> and {NewOperatorRefreshToken}
    ExpO->>ExpO: Save new {NewOperatorRefreshToken}
    ExpO->>BE: 200 OK <br> {NewOperatorAccessToken}<br>{NewOpeatorRefreshToken}
  end 
end
alt Regular Access to API
  BE->>ExpO: Access Operator CAMARA API <br> Authorization: Bearer {NewOperatorAccessToken}        
  ExpO->>ExpO: Decrypts NewOperatorAccessToken,<br>grants Access,<br>progresses request to API Backend,<br>gets API response  
  ExpO->>BE: CAMARA API Response
  Note over BE,FE: Response
end
```

**Key concepts**

- Same considerations that apply for general call flows in [Authorization flows / grant types](#authorization-flows--grant-types) also apply for this flow.

**Flow description**

When a network feature is needed (step 1), the API invoker tries to consume a CAMARA API with an Access Token, but it has already expired, so it gets an error (steps 2-3). 

> Upon this error, the API invoker performs standard Refresh Token Grant flow, providing the Operator Refresh Token in order to get a new pair of Operator Refresh Token and Operator Access Token (step 4). 

Reviewer 2: ```suggestion
Upon this error, the API invoker performs standard Refresh Token Grant flow, providing the Refresh Token in order to get a new pair of Refresh Token and Access Token (step 4). 
```

> The Operator validates the refresh token and retrieves the related information, i.e.: scopes/purpose (step 5) and checks the legal basis of the purpose associated with the token (step 6). If the purpose requires the user's consent, the Operator validates whether the consent is granted (step 7). If the consent is NOT granted e.g., the consent has expired or the user has revoked the consent, the Operator will not issue a new token and will return an error (step 8) and the refresh token flow will fail. 

Reviewer 2: ```suggestion
The Authorization Server validates the refresh token and retrieves the related information, i.e.: scopes/purpose (step 5) and checks the legal basis of the purpose associated with the token (step 6). If the purpose requires the user's consent, the Operator validates whether the consent is granted (step 7). If the consent is NOT granted e.g., the consent has expired or the user has revoked the consent, the Authorization Server will not issue a new token and will return an error (step 8) and the refresh token flow will fail. 
```

> If the user's consent is granted or not required for the applied legal basis, the Operator generates a new pair of Operator Access Token and Operator Refresh Token (step 9), saves the refresh token (step 10), and responds to the API invoker with this information (step 11). 

Reviewer 2: ```suggestion
If the user's consent is granted or not required for the applied legal basis (purpose), the Authorization Server generates a new pair of Access Token and Refresh Token (step 9), saves the refresh token (step 10), and responds to the API invoker with this information (step 11). 
```

> If the user's consent is granted or not required for the applied legal basis, the Operator generates a new pair of Operator Access Token and Aggregator Access Token (step 9), saves the refresh token (step 10), and responds to the API invoker with this information (step 11). 

Reviewer 2: This text can't be added because it adds `Aggregator Access Token` and `Operator Access Token` which are unknown concepts in Camara.

Author: @AxelNennker My bad, one of the things I did before moving the documentation was to remove any concept unknown to CAMARA like the aggregation model of GSMA Opengateway. Good catch, thanks. I will review the PR to correct it.

Author: Typo fixed in https://github.com/camaraproject/IdentityAndConsentManagement/pull/123/commits/354d61853ede084cc57d42e14e4affea02ad5078.
